### ISSUE TITLE
Selectively opting in to `exactOptionalPropertyTypes`

### ISSUE THREAD

**Issue by joshkel** [microsoft/TypeScript#48086]
# Suggestion

I have one or two data types where misuses of `undefined` for optional members have caused multiple bugs. I'd love to be able to update my type definitions and TypeScript configuration to catch these mistakes. `exactOptionalPropertyTypes` does what I want; however, enabling it also raises hundreds of compilation errors throughout my project, in areas of the code where the use of `undefined` is just fine.

Is there, or could there be, a way to enable `exactOptionalPropertyTypes` for individual types? Maybe something like the following?

```ts
interface SampleType {
  a?: string | undefined;
  b?: string;
  c?: string!;
}

let value: SampleType = {};
value.a = undefined; // always legal
value.b = undefined; // legal if exactOptionalPropertyTypes is false
value.c = undefined; // never legal
```

## 🔍 Search Terms

exactOptionalPropertyTypes

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [X] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [X] This wouldn't change the runtime behavior of existing JavaScript code
* [X] This could be implemented without emitting different JS based on the types of the expressions
* [X] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [X] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by flying-sheep** [comment]
Yeah, that would be great, especially since it seems to be buggy in some cases?

**Comment by haukepribnow** [comment]
Here's a link to one of my questions on Stack Overflow with code that could benefit from a feature like this: https://stackoverflow.com/questions/76358299/how-to-define-type-of-object-that-does-not-have-a-specific-field-even-not-set-t

**Comment by aredridel** [comment]
This would be so useful since it affects the local interpretation, not global; it'd be possible to opt in file by file. It would mean carrying `x?: string` vs `x?: string | undefined` around so that code can be both ways.

**Comment by cainlevy** [comment]
I would appreciate this or something similar that allows gradual adoption of `exactOptionalPropertyTypes`. Currently we rely on major libraries which are not ready, and the workarounds are so undesirable that we would not come out ahead trying to adopt this for our own code.
